### PR TITLE
feat: configurable chain config

### LINKED ISSUE #240: Configurable chain config (duh)
### Is there an existing issue?

- [X] I have searched the existing issues

### Motivation

Last week with @Mohiiit we talked about where to put new configuration params - sometimes the line between what should be put in `ChainConfig` vs what should be put in cli arguments is a bit blurry
Let's fix that!

### Request

I believe the difference should be related to how the user uses madara, not about technical details or whatnot.

CLI args should be how node operators (people running madara nodes) configure madara, and chain config should be how appchain developers (kakarot pragma & cie) configure their chain.
This means that block times, and other details like pending block tick time should be in chain config - even though it may be a debatable. The batch size for block production should also be in chain config instead of cli arguments as it is very related to pending block tick time -- we should expect people to tune both at the same time. This may seem a bit strange because this is a performance setting though.

In addition,
- Chain config should be configurable from a .yml (or .toml?) file that we pass as argument `--chain-config ./chain.yml`.
- The docs should provide an annotated (with comments) example file
- Sometimes we would like to override some of the chain config via cli args -- for example, to set a special chain_id for a devnet or e2e test. It would look like `--chain-config-override chain_id=MADARA_TEST`.
  - technical details: this is easy to do, we would just need to load the chain config as a `serde_yaml::Value`, override the value we want by writing fields into that untyped `Value`, and then use `serde_yaml::from_value` to get a ChainConfig struct. This means that this would reuse the same code for deserializing Felts and stuff as the config file, the deserialization is done by serde.
  - stuff like versioned_constants may require more thought to deserialize from cli args - let's just not support that because I don't really see any usecase for that

Plus,
- We may want "presets" in the future, to configure in a single cli arg a devnet/full node syncing from mainnet/other chains
  - i don't know how this would really look in practice, maybe we can get away with just switching between different preset chain configs? the question is whether presets should affect cli arg config.

Thoughts? :)

I am not sure if yml is better than toml, I think starknet people are more familiar with toml? what do you think?
I don't want json as it can't have comments.

### Solution

_No response_

### Are you willing to help with this request?

Yes!

### PR DESCRIPTION
**Pull request by azurwastaken**
# Pull Request type

- Feature

## What is the current behavior?

Resolves: #240 

## What is the new behavior?

- We can now provide a yaml with chain config description or use presets to populate chain config struct

## Does this introduce a breaking change?

No

### CODE REVIEW

File: config.yaml.example [hunk 1]
```diff
@@ -0,0 +1,33 @@
+chain_name: "Starknet Mainnet"
```
Comment (antiyro) [inline config.yaml.example]: Could you drop some documentation describing each of these fields?

Comment (azurwastaken) [reply config.yaml.example]: Done

File: crates/client/eth/src/l1_messaging.rs [hunk 2]
```diff
@@ -315,6 +315,21 @@ mod l1_messaging_tests {
         }
     );
 
+    #[fixture]
```
Comment (Trantorian1) [inline crates/client/eth/src/l1_messaging.rs]: Couldn't this be refactored to be more reusable given your are using the same code in multiple places?

File: crates/node/src/cli/chain_config_overrides.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    #[arg(long, requires = "chain_config_override", value_name = "OVERRIDED MAX NONCE VALIDATION FOR SKIP")]
+    pub max_nonce_for_validation_skip: Option<u64>,
+    #[arg(long, requires = "chain_config_override", value_name = "OVERRIDED ETH CORE CONTRACT")]
+    pub eth_core_contract_address: Option<String>,
+}
+
+impl ChainConfigOverrideParams {
+    pub fn override_cfg(&self, mut chain_config: ChainConfig) -> ChainConfig {
+        let params = self.clone();
+        (
```
Comment (Trantorian1) [inline crates/node/src/cli/chain_config_overrides.rs]: This is pretty hard to decypher, consider just assigning the variables individually

File: crates/node/src/cli/mod.rs [hunk 4]
```diff
@@ -76,6 +97,21 @@ impl RunCmd {
         self.name.as_ref().expect("Name was just set")
     }
 
+    pub fn get_config(&self) -> anyhow::Result<Arc<ChainConfig>> {
+        let mut chain_config: ChainConfig = match &self.preset {
+            Some(preset_name) => ChainConfig::from_preset(preset_name.as_str())?,
+            None => {
+                ChainConfig::from_yaml(&self.chain_config_path.clone().expect("Failed to retrieve chain config path"))?
```
Comment (Trantorian1) [inline crates/node/src/cli/mod.rs:104]: Maybe a more explicit error for when a chain config has not been specified at all and no preset name was given?

**Review by Trantorian1 — COMMENTED** [review]
All seems good. Check comments for a few changes to change, maily refactoring.

File: crates/node/src/cli/mod.rs (hunk 4 above)
Comment (azurwastaken) [reply crates/node/src/cli/mod.rs:104]: if neither a yaml or preset is given, it throw an error due to #[clap(group(ArgGroup::new("chain_config").args(["chain_config_path", "preset"]).required(true)))]

File: crates/client/eth/src/l1_messaging.rs (hunk 2 above)
Comment (azurwastaken) [reply crates/client/eth/src/l1_messaging.rs]: created a test_common in the utils crate crates/primitives/utils/src/tests_common.rs

File: crates/node/src/cli/chain_config_overrides.rs (hunk 3 above)
Comment (azurwastaken) [reply crates/node/src/cli/chain_config_overrides.rs]: changed

**Review by EvolveArt — APPROVED** [review]

**Review by Trantorian1 — APPROVED** [review]
All good :+1:

**Pull request merged**